Deprecate estimate method in favor of class constructor

matthew-brett opened the pull request:
## Description

This is the promised follow-on PR from the discussion more or less starting at https://github.com/scikit-image/scikit-image/pull/7754#issuecomment-2742974392.

To recap briefly:

There is a wart in the current API, in that the `estimate` method of Transforms ignores any stored state of the initialized Transform, and re-estimates.   So, for example:

```python
tf = SimilarityTransform(rotation=(0.2, 0.3))
tf.estimate(src, dst)
```

wil discard the initial state with the specified rotations.   Another problem is that one generally has to first initialize the transform, most sensibly as an identity transform, and then estimate, thus:

```python
tf = SimilarityTransform.identity(3)
tf.estimate(src, dst)
```

However, the argument to the identity transform is redundant, because we can infer the dimensionality of the points (3) from `src` and `dst` point arrays.    So, this PR deprecates the `estimate` method, in favor of a class constructor, like this:

```python
tf = SimilarityTransform.from_estimate(src, pts)
```

### For discussion

* We floated various names for the class constructor.  `from_estimate` gained favor from me (MB) (I proposed it) and @stefanv.  @jni leaned towards some fancy footwork to allow `estimate` to be *both* the original estimation method, *and* the class constructor, using the [trick here](https://stackoverflow.com/questions/20533349/python-regular-method-and-static-method-with-same-name/28766809#28766809).   I [thought that was going to be confusing](https://github.com/scikit-image/scikit-image/pull/7754#issuecomment-2748623196).  We can rehearse that argument more.
* In writing the tests, I (MB) discovered an error in the check for a degenerate tranform, [discussed here](https://github.com/scikit-image/scikit-image/issues/7770).   I used an obvious but quite possibly wrong workaround (mentioned in the linked issue).  I would be happy to be corrected.
* One wrinkle is that the `from_estimate` API returns `None` for failed estimation.   This is generally just as informative as the `estimate` method, except in the case of `PiecewiseAffineTransform`, where the estimation involves multiple steps, and resulting affines.   The `estimate` method collects an affine for each step, recording an all-NaN affine for steps that fail estimation, so you can track, from the resulting `tform.affines` attribute, where the estimation failed.   The returned `None` does not give you that granular information.  Is there a use for that information?  How should it reach the user?
* Should we also adapt the API-a-like classes `LineModelND`, `CircleModel`, `EllipseModel` at https://github.com/scikit-image/scikit-image/blob/main/skimage/measure/fit.py#L27 to have a `from_estimate` class method (instead of an `esimate` method)?
* Similarly, should we also adapt the `ThinPlateSpline` class (https://github.com/scikit-image/scikit-image/blob/main/skimage/transform/_thin_plate_splines.py#L7)?  (I think so).

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
In `skimage.measure`, add a new class method and constructor `from_estimate` for 
`LineModelND`, `CircleModel`, and `EllipseModel`. This replaces the old
API—the now deprecated `estimate` method—which required initalizing a model
with undefined parameters before calling `estimate`. 
```

```release-note
In `skimage.transform`, add a new class method and constructor `from_estimate` for
`AffineTransform`, `EssentialMatrixTransform`, `EuclideanTransform`,
`FundamentalMatrixTransform`, `PiecewiseAffineTransform`, `PolynomialTransform`, 
`ProjectiveTransform`, `SimilarityTransform`, and `ThinPlateSplineTransform`. This
replaces the old API—the now deprecated `estimate` method—which required initializing an undefined transform before calling `estimate`.
```

matthew-brett commented:
This one is ready for a first review - but we need to make some decisions, listed above, so I would be very glad of some initial feedback.

matthew-brett commented:
I'll proceed with these (provisional) answers to the points for discussion above:

* `from_estimate` is a reasonable name for the class constructor method.    Conversely, using `estimate` will commit us to using a odd trick that allows the class method to differ from the instance method - contrary to the normal interface in Python - and for the long term (because we will have to keep an error marker for the instance method).
* I believe my check for the degenerate transform is (now) correct (directly evaluate the determinant of the generated transform matrix).
* I guess that the use-case is rare, of investigating the various affines in a `PiecewiseAffineTransform`, and that rare user that needs to investigate, can read the code for themselves to reproduce the steps.  So `None` is OK as a return value from `from_estimate`.   If it proves necessary to get the component affines, we can add that as another class method, I guess.  But not in this PR, as likely [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).
* I will implement `from_estimate` and `estimate` deprecations for `LineModelND`, `CircleModel`, `EllipseModel` and`ThinPlateSpline`.

When I've done the last step, I'll unmark this PR as draft.

If anyone disagrees with the above - let me know soon?

stefanv commented:
That all sounds good to me. I'm fine with `from_estimate`. I do think we need a way to tell the user whether the transform succeeded or failed in all cases. Raising may be a fine option, unless there is some reason to proceed with a non-fitted model.

stefanv commented:
Nevermind, I had misunderstood what you wrote. I guess the question is: do we return None on failure, or error out? It may be unexpected to get a non-tf-compliant result without obvious failure.

matthew-brett commented:
I like that `from_estimate` conveys that we are making a transform from an estimate - it has a classic class constructor feel.   Whereas `try_estimate` feels like it will trigger some confusion - is this like a `try ... except`?

I have a vague positivity about the current API - in that there is no way that someone could accidentally use a `None` returned Transform.   For the option of raising an error - we really aren't saving them that much code.  I mean comparing:

```python
tf = AffineTransform.from_estimate(src, dst, raise_on_failure=True)
```

to:

```python
tf = AffineTransform(src, dst):
if tf is None: raise ValueError('Estimation failed')
```

lagru reviewed (commented):
I gave it a first look. A lot of the stuff is just nitpicks – please tell me if it doesn't feel constructive and I should tone it down. :wink: But I'd like to get your input on a few items.

stefanv commented:
Matthew, re: the None return: it's not that it's hard to handle, more that it hides failures in a way that may not be obvious to developers. E.g., if you run a program and it works fine the first ten times, but then the 11th time breaks because None doesn't have some attribute, it's going to be confusing. If the error is "we couldn't fit the model", you immediately know what went wrong.

Re: naming, I prefer `from_estimate`, since that's already a paradigm in pandas etc.

lagru commented:
Re https://github.com/scikit-image/scikit-image/pull/7771#issuecomment-2793811560, I can live happily with `from_estimate`. :smile: 

Regarding raising on failure, this would be a way to tell users why an estimation failed. Just returning `None` has a "computer says no" vibe. :sweat_smile: But we could also work with warnings instead. Or add optional and more specific exceptions later. So I'm happy to keep this out of the scope of this PR.

stefanv commented:
As always with API design, I just have a vague feeling, so I'll have to rely on y'all to help me figure out if the concern is legit or not.

lagru commented:
Could we perhaps use a nice sentinel value other than `None`? That should result in nice error messages like "<FailedEstimation> has no attribute...". I would also provide a nice container for a string giving a reason for the failure. Though, we would need a good to check this maybe with something like
```python
class FailedEstimation:
    def __bool__(self):
        return False
```
that would still allow a simple `if not tform:` check or `assert tform`.

matthew-brett commented:
> Could we perhaps use a nice sentinel value other than `None`? That should result in nice error messages like "<FailedEstimation> has no attribute...". I would also provide a nice container for a string giving a reason for the failure. Though, we would need a good to check this maybe with something like
> 
> ```python
> class FailedEstimation:
>     def __bool__(self):
>         return False
> ```
> 
> that would still allow a simple `if not tform:` check or `assert tform`.

What would be the advantage over raising an error?

lagru commented:
> What would be the advantage over raising an error?

As stated above, it's easier to handle. `try...except` forces users to lookup and match the correct exception. But perhaps still better than a weird custom object that is "falsy".

matthew-brett commented:
> > What would be the advantage over raising an error?
> 
> As stated above, it's easier to handle. `try...except` forces users to lookup and match the correct exception. But perhaps still better than a weird custom object that is "falsy".

Ah yes - sorry - I see the point.   I guess at the moment, my order of preference is:

* falsey return value (@lagru's suggestion).  I guess we could add an error message there, perhaps returned by `__str__`.
* raise an error in class constructor.
* None return value.

but I don't feel very strongly.   What do y'all think?

matthew-brett reviewed (commented):
Some comments on comments.  Nitpickey is fine ...

matthew-brett commented on skimage/transform/tests/test_geometric.py:
```diff
@@ -95,18 +95,28 @@ def test_euclidean_estimation():
     assert_almost_equal(tform3.params, tform2.params)
+    # via estimate method
+    tform4 = EuclideanTransform.identity()
+    with pytest.warns(FutureWarning, match='`estimate` is deprecated'):
+        assert tform4.estimate(SRC, DST)
```
Yes, go for it.

matthew-brett replied:
Should we mark this as resolved?

matthew-brett commented on skimage/transform/_geometric.py:
```diff
@@ -1188,19 +1379,22 @@ def estimate(self, src, dst):
         self._tesselation = spatial.Delaunay(src)
 
         success = True
+        null_params = np.full((d + 1, d + 1), np.nan)
```
No - it's the same case - will rename for clarity.

matthew-brett reviewed (commented):
A couple of comments - thanks.

matthew-brett commented:
Returning to the API question ... it seems to me that one will often want to do this, as @lagru points out:

```python
tf = Transform.from_estimate(src, dst)
if not tf:
    "Respond in some way"
```

So far, the None and `FailedEstimation` ideas both serve.  The FailedEstimation improves on None in allowing - say - a `.message` to give more information, at the expense of being a bit more fancy and less obvious.

Otherwise (the `raise` option), one will nearly always have to do something like:

```python
try:
    tf = Transform.from_estimate(src, dst)
except TransformEstimationError:
    "Respond in some way"
```

Of course, we could give the user a choice with:

```python
 tf = Transform.from_estimate(src, dst, raise=True)
```

where the user then passes on the problem to their user.  But this seems like an example where "There should be one-- and preferably only one --obvious way to do it."   And we aren't saving the user much compared to:

```python
tf = Transform.from_estimate(src, dst)
if not tf:
    "Respond in some way"
```

The advantage of Lars' FailedEstimation is that the user can even do:

```python
tf = Transform.from_estimate(src, dst)
if not tf:
    raise ErrorOfYourChoice(str(tf))
```

So I'm still vaguely in favor of Lars' `FailedEstimation` idea.   Any more thoughts?

matthew-brett commented:
As a side-point, and returning to [this item](https://github.com/scikit-image/scikit-image/pull/7771#issuecomment-2792184637):

> * I guess that the use-case is rare, of investigating the various affines in a `PiecewiseAffineTransform`, and that rare user that needs to investigate, can read the code for themselves to reproduce the steps.  So `None` is OK as a return value from `from_estimate`.   If it proves necessary to get the component affines, we can add that as another class method, I guess.  But not in this PR, as likely [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Lar's `FailedEstimate` allows us a simple way to return the intermediate results of the failed estimation, if we want to do that.

matthew-brett commented:
@lagru - you mentioned resolving the stacklevel for the warnings.   I wonder whether you could help me with some confusing stacklevel test errors?   Example at https://github.com/matthew-brett/scikit-image/actions/runs/14408547193/job/40410880036#step:9:445:

```
        warning_message = (
            "Standard deviation of data is too small to estimate "
            "circle with meaningful precision."
        )
        with pytest.warns(RuntimeWarning, match=warning_message) as _warnings:
            assert CircleModel.from_estimate(np.ones((6, 2))) is None
>       assert_stacklevel(_warnings)
E       AssertionError: Warning with wrong stacklevel:
E         Expected: /opt/hostedtoolcache/Python/3.13.2/x64/lib/python3.13/site-packages/skimage/measure/tests/test_fit.py:190
E         Actual: /opt/hostedtoolcache/Python/3.13.2/x64/lib/python3.13/site-packages/skimage/_shared/utils.py:572
E         RuntimeWarning: Standard deviation of data is too small to estimate circle with meaningful precision.
```

This appears to be because I'm using the now-deprecated `estimate` method from the `from_estimate` class method, and ignoring the warning on the way:

```python
    @classmethod
    def from_estimate(cls, data):
        instance = cls()
        with catch_warnings():
            filterwarnings("ignore", message="`estimate` is deprecated")
            success = instance.estimate(data)
        return instance if success else None
```

but I must admit I didn't understand the stacklevel code in `_shared/utils.py`.  Do you see what the problem is?

matthew-brett commented:
OK - work now basically done, waiting for a decision on the return value from the `from_estimate` class methods.   Any final words?   Otherwise I'll implement @lagru's suggestion.

stefanv commented:
I would look at other `from_` constructors in the ecosystem and try to align with the most common pattern.

The FailedEstimation notion feels pretty unusual to me, in terms of typing: `Transform | FailedEstimation`, so my preference is still raising, but I won't hold things up, especially if there are examples of other libraries doing the same.

What happens in scipy.optimize.minimize, which returns a compound class, or in pandas.from_dict, e.g.?

matthew-brett commented:
Team - I need help on this issue : https://github.com/scikit-image/scikit-image/issues/7770

The issue is that the test for the degenerate matrix in `ProjectiveTransform.estimate` was only passing because of a bug in the test (estimating 3D transform starting with a 2D transform).

The obvious fix is to check directly for rank deficiency: https://github.com/scikit-image/scikit-image/pull/7771/files#diff-34acab621bfddc46a8c62f72a08284cd20d06058b76682476a2bc37a11f80b24R1036

However, this generates a new error, revealed in the docs build, but now tested here, with code from the relevant example:

https://github.com/scikit-image/scikit-image/pull/7771/files#diff-34acab621bfddc46a8c62f72a08284cd20d06058b76682476a2bc37a11f80b24R1228

Here, with the new check for rank deficiency, there are multiple failed estimations for the *inverse* affines.  Specifically, the estimated inverse affine for the the first point is:

```python
<AffineTransform(matrix=
    [[ 1.00000000e+00, -1.34103013e-15,  8.52651283e-13],
     [ 0.00000000e+00,  0.00000000e+00,  5.12000000e+02],
     [ 0.00000000e+00,  0.00000000e+00,  1.00000000e+00]]) at 0x10b490f80>
```

that has a determinant of nearly exactly 0, and singular values of `[5.12000977e+02, 1.00000000e+00, 1.87047690e-19]`.

So - what do you think about the strict check?  Should we just drop it?   Should we note that the inverse is inestimable, but nevertheless return the forward transform?

matthew-brett commented:
Ok - for now I've removed the more strict check for degenerate transform matrices, and I've removed the offending test (for https://github.com/scikit-image/scikit-image/pull/6207) - because it cannot succeed without the stricter check.  I've moved the intended purpose of the removed test to another place (L950 in `test_geometric.py`).

But I think we do have to decide what to do about these degenerate matrices.

matthew-brett commented:
In reply to @stefanv's question on what other libraries do:  I'll stick to estimates, rather than incompatible data, etc.

## Statsmodels

Statsmodels generally seems to warn rather than raising for failed estimation.  For example, consider this code:

```
import numpy as np
import statsmodels.api as sma

groups = np.repeat([0, 1], [20, 20])
# Perfect separation.
data = np.random.normal(size=40) + groups * 30
# Design matrix.
X = np.column_stack((data, np.ones(40)))
sma.Logit(endog=groups, exog=X).fit().summary()
```

This leads to warnings of form:

```
/Users/mb312/Library/Python/3.12/lib/python/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
Warning: Maximum number of iterations has been exceeded.
         Current function value: 0.000000
         Iterations: 35
/Users/mb312/Library/Python/3.12/lib/python/site-packages/statsmodels/base/model.py:607: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
```

You have to tell Statsmodels explicitly that you want an error in this case:

```python
model = sma.Logit(endog=groups, exog=X)
model.raise_on_perfect_prediction = True
# This raises a PerfectSeparationError
model.fit()
```

## Scipy.optimize.minimize

Here's an example:

```python

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 32.43954058587445
        x: [ 3.013e+01]
      nit: 4
      jac: [ 4.768e-07]
 hess_inv: [[ 2.500e-02]]
     nfev: 12
     njev: 6
```

Notice the `success` flag, that [can be False](https://stackoverflow.com/questions/62457289/scipy-optimize-minimize-not-converging-giving-success-false) (but should be True in this case).

However, I can set `bounds` outside the correct value, and I get this:

```python
ins] In [22]: sp.optimize.minimize(cost, 0, bounds=[[10, 20]])
Out[22]: 
  message: CONVERGENCE: NORM OF PROJECTED GRADIENT <= PGTOL
  success: True
   status: 0
      fun: 2085.416683663995
        x: [ 2.000e+01]
      nit: 1
      jac: [-4.053e+02]
     nfev: 4
     njev: 2
 hess_inv: <1x1 LbfgsInvHessProduct with dtype=float64>
```

Notice `success` is still true, I have to look at the message to see there is a problem.  I don't think `scipy.optimize.minimize` gives an error on normal use.

## Scikit-learn

See : https://github.com/scikit-learn/scikit-learn/blob/98ed9dc73/sklearn/exceptions.py

They appear never to raise an error on estimation (as opposed to invalid data dimensions etc).

## Pandas

I don't think Pandas is ever doing estimation in its `from_*` functions - rather - it is checking data for compatibility with the rules, and therefore, something like `ValueError` is sensible.

## Summary

The general preference seems to be to warn, rather than raise, and return the best effort result. 

Up until now, we have been:

* Not warning
* Requiring users to check explicitly if the transform estimation succeeded with `assert tf.estimate(src, dst)` or similar.
* Returning an all `np.nan` transform in `tf.params` - that the user can try and use, but will eventually discover is not sensible.

If we raise an error, we're going to force a lot of `try ... except` blocks in user code.  I personally don't like these very much, but that's a personal preference.

It's odd, true, to return this unusable `EstimationFailure` result thing, but actually, it seems more sensible than what we are doing now (it's easy to miss the failed estimation, and the all `np.nan` result would be a little confusing to the unwary) - and I don't think `EstimationFailure` would be all that confusing - the user will see what's going on as soon as they try and use it.  I mean with something like:

```python
class EstimationError(ValueError):
    """ Error for failed estimation """


class FailedEstimation:

    def __init__(self, message):
        self.message = message

    def __bool__(self):
        return False

    def __str__(self):
        return self.message

    def __call__(self, *args, **kwargs):
        raise EstimationError(f'Call on failed estimation: {self.message}')

    def residuals(self, *args, **kwargs):
        raise EstimationError(
            f'residuals on failed estimation: {self.message}')

    @property
    def params(self):
        raise EstimationError(f'inverse for failed estimation: {self.message}')

    @property
    def inverse(self):
        raise EstimationError(f'inverse for failed estimation: {self.message}')
```

stefanv commented:
But, do we now expect the user to use isinstance to figure out whether the estimate worked? That's seems like a way more advanced Python feature that many users of functional APIs may not even be aware of?

stefanv commented:
(In none of those examples does the library change output type based on success.)

matthew-brett commented:
Oh - no - the idea is they use truthiness - in a close parallel to what they are doing now.  As in:

```python
tf = Transform.from_estimate(src, dst)
if not tf:
      """ Do whatever """
```

or:

```python
if not (tf := Transform.from_estimate(src, dst)):
    raise ValueError(f'Estimation failed: {tf}')
```

I don't think it matters all that much whether the output type changes - as long as it has a matching API.   I mean - None is a different output type - isn't that a fairly common pattern?

lagru reviewed (commented):
@matthew-brett, thanks for the sleuthing and summary of how other libraries deal with failed estimation. That somewhat confirms my feeling that a failed estimation is an expected outcome for which raising an exception doesn't feel right. 

After digging a bit into the stacklevel related failures, I took the liberty to commit a few changes to your PR since that seem a lot easier than explaining. I applied a few – I expect non-controversial – tweaks too. Still please have a look and feel welcome to discuss or modify anything. :)

lagru commented on skimage/measure/fit.py:
```diff
@@ -680,6 +747,30 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
+    warn(
+        "Passing custom classes without `from_estimate` deprecated "
+        "since version 0.26 and will be removed in version 2.2. "
+        "Add `from_estimate` method to custom class to avoid this warning.",
```
Hmm, this doesn't mention that we expect and call `from_estimate` as a classmethod. And asking users to implement a classmethod seems a bit more advanced than asking for a simple method. 

I'm wondering if we can provide an abstract base or – long-term – an [protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) instead. That would also make the documentation a lot easier since we could simply point them at that class.

matthew-brett replied:
I've noted that `from_estimate` should be a class method.   For the protocol - yes - we could do that - but given in this case we only need two things (`from_estimate` and `residuals`), and given that making a custom class will be somewhat advanced, perhaps we can just extend the docstring a bit to improve the documentation?

lagru replied:
I don't follow? Providing an ABC, pointing at it and saying "hey this is the API we expect from `model_class=`" seems a lot simpler than explaining that in a docstring?

But I'm okay with giving the docstring a shot. :+1:

matthew-brett replied:
I'm not against an ABC - only saying that once you've got to the level of the kind of user that would benefit from an ABC for this two-method API, they probably won't be put off by the instruction to define a class method.

matthew-brett replied:
OK - I added a `Protocol`.  It turns out you can't (as far as I can see) use Protocol to require a method to be a class method, so I've added a separate check for that, and added tests for protocol violations.

lagru commented on skimage/measure/fit.py:
```diff
@@ -974,4 +993,5 @@ def ransac(
 
-    return model, best_inliers
+    # Return model from wrapper, otherwise model itself.
+    return getattr(model, 'model', model), best_inliers
```
Continued discussion in https://github.com/scikit-image/scikit-image/pull/7771/files#r2042927190.

matthew-brett commented on skimage/transform/_geometric.py:
```diff
@@ -498,6 +532,32 @@ def _setup_constraint_matrix(self, src, dst):
+        Returns
+        -------
+        tf : :class:`FundamentalMatrixTransform` instance
+            Transform estimated from `src` and `dst`
```
I changed the text to say that the algorithm needs 8 points.  It will error with fewer than 8 points, so the current docstring was a bit confusing.

matthew-brett commented on skimage/measure/fit.py:
```diff
@@ -680,6 +747,30 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
+        return cls
+
+    warn(
+        "Passing custom classes without `from_estimate` deprecated "
```
I've added this in another commit.

matthew-brett reviewed (commented):
Some comments on kind review.

lagru commented on skimage/_shared/utils.py:
```diff
@@ -402,6 +435,90 @@ def _docstring_add_deprecated(func, kwarg_mapping, deprecated_version):
+    Fetch the original (not-wrapped) method before applying this decoration.
+    """
+
+    def estimate(self, *args, **kwargs):
+        return self._estimate(*args, **kwargs)
```
Re the current test failures, I think the new approach relying on a wrapped closure is mucking up the stacklevel detection. I'm guessing that doesn't work for closures. 

Is there a reason you chose to remove the actual `estimate` method in favor of adding it back with this decorator? Seems more complicated and indirect? I'm happy to look into a solution but wanted to check the line of reasoning.

If I understand correctly, the current approach copies docstring on `from_estimate` to `estimate`. That isn't ideal because `from_estimate` returns something different than `estimate`.

matthew-brett replied:
I was trying to solve three problems; the first was the stacklevel matching, the second was the inconvenience of having to use a deprecated function inside a not-deprecated one (using `estimate` from within `from_estimate`, with the associated warnings ignore machinery), and the third was the annoyance of having to have a very similar copy of the docstring in both `from_estimate` and `estimate`.  This way solves those problems, and in particular, the decorator that generates `estimate` replaces the `Returns` section in `from_estimate` docstring with the relevant `Returns` section for `estimate`.

Is the new stacklevel code correct and useful for other cases?   These edits mean the tests can pass with the original stacklevel code - but is the new stacklevel code preferable for the future?

matthew-brett replied:
I still moderately like the `_estimate` solution for reasons given above.  Yes, it's magic, but a magic the user will not see.    Can we mark this as resolved?  Or should we discuss more?

lagru replied:
But on second thought, I don't want to make you do that additional work. It's probably not worth it.

lagru commented on skimage/_shared/utils.py:
```diff
@@ -402,6 +402,40 @@ def _docstring_add_deprecated(func, kwarg_mapping, deprecated_version):
+
+    Notes
+    -----
+    The ``estimate`` method may already be wrapped — for example, if we have
+    inherited from a class to which this decorator has already been applied.
+    Fetch the original (not-wrapped) method before applying this decoration.
```
I wouldn't bet on me catching that. :sweat_smile: But let's first settle on a general approach and then iron out kinks like this.

lagru commented on skimage/_shared/utils.py:
```diff
@@ -402,6 +431,96 @@ def _docstring_add_deprecated(func, kwarg_mapping, deprecated_version):
+    )
+
+
+def _replace_estimate_returns(docstring):
```
Just noticed, this function is defined twice? See directly above?

matthew-brett commented on skimage/measure/tests/test_fit.py:
```diff
@@ -157,20 +175,19 @@ def test_circle_model_residuals():
         "circle with meaningful precision."
     )
     with pytest.warns(RuntimeWarning, match=warning_message) as _warnings:
-        assert not model.estimate(np.ones((6, 2)))
+        assert CircleModel.from_estimate(np.ones((6, 2))) is None
     assert_stacklevel(_warnings)
```
I think you've fixed that now with the manual stacklevel?

lagru commented:
Yes, exactly like that. I have tested this only somewhat, but something like this should get close to the intended behavior?

```python
class FailedEstimationMixin:
    @classmethod
    def _as_failed_estimation(cls, reason="Estimation failed."):
        """Create failed estimation (alternative constructor)."""
        failed = cls()
        failed._fail_reason = reason
        return failed

    def __getattribute__(self, item):
        if (
            not item.startswith("_")  # only raise on public attributes / methods
            and "_fail_reason" in object.__getattribute__(self, "__dict__")
        ):
            reason =  object.__getattribute__(self, "_fail_reason")
            msg = (
                f"Cannot access {item} of failed estimation. {reason}\n\n"
                f"You can check for a failed estimation with\n\n"
                f"    if estimation:\n"
                f"        # handle successful estimation"
            )
            raise AttributeError(msg)
        return object.__getattribute__(self, item)

    def __repr__(self):
        if "_fail_reason" in object.__getattribute__(self, "__dict__"):
            reason =  object.__getattribute__(self, "_fail_reason")
            out = f"<{type(self).__qualname__}: Failed estimation ({reason})>"
            return out
        return super().__repr__()

    def __bool__(self):
        return "_fail_reason" not in object.__getattribute__(self, "__dict__")
```

You'd use `_as_failed_estimation` in the `from_estimation` constructor to return a modified version.

This approach seems a bit more complex than I initially believed. And there's the outstanding problem that an existing `__repr__` would conflict with the one from the same inheritance level of the mixin. But it's at least a draft to help evaluate this option.

We could also opt for a decorator that modifies each estimation class but that would have its own disadvantages.

Probably best if we let other core devs chime in on a central decision API like this.

matthew-brett commented:
Your suggestion is that we have two types of instances of any given Transform class - a successful estimation version (as we have now) and a failed estimation version, where we disable most of the methods and attributes.

That does solve the potential confusion of returning different types of instances depending on the success of the estimation - at the expense of - I think - considerably complexity in the Transform object API.      I also don't think it's a major problem that `from_estimate` can return two possible different types of instance - the `None` example is - I think - fairly common.   But maybe someone has a suggestion for where, in practice, that might matter?

matthew-brett commented:
Note - add link to papers.

jni commented:
Thanks all! The discussion was easy to follow. I started at the bottom with the FailedEstimation class thinking, *What! This is an outrage! Why not use None like a normal human!* 😂 but then I read all the prior discussion and had nothing further to add. 😊

matthew-brett commented:
@lagru - refactoring done now.  I need to add some documentation.   Could I ask for your help on the type annotations?

matthew-brett commented:
Could I ask the advice of the team about the `__init__` functions for `LineModelND`, `CircleModel` and `Ellipsemodel`?

At the moment these all take no parameters and set `self.params = None` - where this is gives a broken transform, that errors on calling `residuals`.

This somewhat made sense when one always, in practice, followed the `__init__` with an `estimate`, but this is no longer the case with the `from_estimate` refactor.

You can see the effect of this in the e.g. `params` argument in a docstring example:

```python
xy = CircleModel().predict_xy(t, params=(2, 3, 4))
```

Surely this should better be:

```python
xy = CircleModel(2, 3, 4).predict_xy(t)
```

or

```python
xy = CircleModel((2, 3), 4).predict_xy(t)
```

I don't think there's any sensible interpretation of an identity for these transforms.

I think the best way for the future is to insist that the user pass sensible values for `self.params` to the constructor, or use `from_estimate`.   As in:

```python
class CircleModel(BaseModel):

    def __init__(self, *args, **kwargs):
        if not args and not kwargs:
            # Raise a deprecation warning.
            self.params = None
            return
        self._check_params(*args, **kwargs)

    def _check_params(self, xc, yc, radius):
        self.params = (xc, yc, radius)

```

But this would take another deprecation cycle for the current e.g. `CircleModel()` calls.   As 

Reasonable to do here?  Or wait for another PR?   Or something else?

matthew-brett commented:
@lagru - I've removed the docstring find / replace - it wasn't going to work for the new examples in the `from_estimate` docstring, because of the possible `FailedEstimation` return.   So I've just copied back the old docstrings and used a more standard deprecator.

jni commented:
> I think the best way for the future is to insist that the user pass sensible values for self.params to the constructor, or use from_estimate. As in:

I agree with this approach. I have no opinion about which PR should handle this.

jni commented:
btw I prefer this form:

```
xy = CircleModel((2, 3), 4).predict_xy(t)
```

We should no longer have APIs that have separate row/col or x/y kwargs — coordinates should be generalisable to nD, even in functions that don't have higher-dimensional implementation, for API consistency with the functions that do.

stefanv commented:
> `xy = CircleModel((2, 3), 4).predict_xy(t)`

@jni So does your comment mean that we should reconsider the `_xy` part?

jni commented:
> So does your comment mean that we should reconsider the _xy part? I.e., just have .predict()?

I don't actually know this part of the codebase, so I don't immediately know what `predict_xy(t)` does… But I think my answer is yes regardless. 😂

matthew-brett commented:
OK - the current state is ready for full review.

lagru commented on skimage/_shared/utils.py:
```diff
@@ -402,6 +440,90 @@ def _docstring_add_deprecated(func, kwarg_mapping, deprecated_version):
     return final_docstring
 
 
+class FailedEstimationAccessError(Exception):
+    """Error from use of failed estimation instance
+
+    This error arises from attempts to use an instance of
+    :class:`FailedEstimation`.
+    """
```
Continuing a chat discussion here which I had to bow out of due to dinner. 

Otherwise, I would have pointed out that[ Python's official docs](https://docs.python.org/3/reference/datamodel.html#object.__getattr__) recommend returning an AttributeError when raising an error in "__getattr__". So this currently deviates from the default behavior. 

@Stéfan van Walt, if understood the discussion correctly you are worried about introducing even more strange concepts. 

I'd argue that returning a simple AttributeError with a descriptive message is the least strange option here. You asked to revert to this which doesn't use an AttributeError. So I don't understand how that fits together.

I'd also like to point out that the whole discussion & ambiguity around where and if to expose `FailedEstimationAccessError` so that users can catch it goes away, if we simply raise a standard AttributeError.

That said, this seems somewhat like a minor issue to me. The error points out how to deal with this – which is using `if ...` instead of `try...except`. So I don't feel particularly obliged to help users use our API in a way we don't intend to.

Also note, we don't really do custom error or warning classes elsewhere in our API.

stefanv replied:
Yes, good.

matthew-brett replied:
Done.

matthew-brett reviewed (commented):
Some comments.

stefanv commented on doc/examples/transform/plot_geometric.py:
```diff
@@ -79,8 +80,8 @@
-# estimate the parameters of a geometric transformation using the least-
-# squares method.
+# generate a transform by estimating the parameters of a geometric
+# transformation using the least- squares method.
```
```suggestion
# transformation using the least squares method.
```

stefanv commented on doc/examples/transform/plot_geometric.py:
```diff
@@ -98,8 +99,44 @@
+# For many transform types, including the ``ProjectiveTransform`, it is
+# possible for the estimation process to fail.  If this is the case,
+# ``from_estimate`` returns a special object of type ``FailedEstimation``.
+# If you apply a truth test (e.g. ``bool(tform3)``, ``FailedEstimation``
```
```suggestion
# If you apply a truth test (e.g. ``bool(tform3)``), ``FailedEstimation``
```

stefanv commented on doc/source/user_guide/geometrical_transform.rst:
```diff
@@ -102,11 +102,10 @@ of points (the source and the destination), as explained in the
-Source and destination points can be determined manually, or using the
-different methods for feature detection available in :mod:`skimage.feature`,
-such as
+The ``from_estimate`` class method uses least-squares optimization to minimize
```
```suggestion
The ``from_estimate`` class method uses least squares optimization to minimize
```

stefanv commented on skimage/_shared/utils.py:
```diff
@@ -34,67 +50,79 @@ def _count_wrappers(func):
 
 def _warning_stacklevel(func):
-    """Find stacklevel for a warning raised from a wrapper around `func`.
+    """Find stacklevel of `func` relative its global representation.
```
```suggestion
    """Find stacklevel of `func` relative to its global representation.
```

stefanv commented on skimage/_shared/utils.py:
```diff
@@ -34,67 +50,79 @@ def _count_wrappers(func):
+    Similar to :func:`count_inner_wrappers`, this counts the number of times
+    `func` has been wrapped. However, this function doesn't start counting
+    from `func` but instead tries to access the "global representation" of
+    `func`. This means that you cold use this function from inside a wrapper
```
```suggestion
    `func`. This means that you could use this function from inside a wrapper
```

stefanv commented on skimage/_shared/utils.py:
```diff
@@ -547,6 +680,59 @@ def wrapped(*args, **kwargs):
+    class_name = func.__qualname__.split('.')[0] if class_name is None else class_name
+    return deprecate_func(
+        deprecated_version="0.26",
+        removed_version="2.2",
```
```suggestion
        removed_version="2.2",
```

Should this be 2.0?

matthew-brett replied:
I was hoping to reduce the need-to-change-now footprint for 2.0.   What do you think?

lagru replied:
It makes probably more sense to schedule this for "2.0 (or later)". That's what I've been proposing in other PRs. It aligns user expectation with the big transition. I think message is "if you deal with deprecations slated for 2.0" you will only have a few breaking changes to worry about. 

We'll likely have to keep pre-v2.0 around on maintenance for quite some time.

matthew-brett replied:
Do we agree we resolved this in the meeting a few days ago?

lagru replied:
I don't exactly remember. I think you advocated for drawing the deprecation out? It feels a bit weird to me to keep this working beyond 2.0. I'd prefer it saying "2.0 (or later)", but I can live with it. We will probably discuss this at least once a gain before the next release. :sweat_smile:

lagru replied:
Note, every deprecation should be noted in our `TODO.txt` so that we remember it. I'll add it.

stefanv replied:
0.28 or 2.2, whichever comes earlier? :thinking: Quite hard to use post-2.0 numbers, since we don't know in how many versions 2.0 will be tagged.

stefanv commented on skimage/_shared/utils.py:
```diff
@@ -547,6 +680,59 @@ def wrapped(*args, **kwargs):
+    return cls
+
+
+def _fix_inherited_from_estimate(cls):
```
Not sure I understand what's happening here, but trust that it's necessary :)

matthew-brett replied:
I'm afraid so - I did go through various iterations on this.   Can unpack if you're interested.

matthew-brett replied:
I'll assume I can resolve this issue.

stefanv replied:
Is this just to get the name to appear correctly? If we can just expand "fix" to be a bit more explicit, that'd be enough.

matthew-brett replied:
Done.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -215,6 +261,25 @@ def predict_y(self, x, params=None):
+        Parameters
+        ----------
+        data : (N, dim) array
+            N points in a space of dimensionality dim >= 2.
```
```suggestion
            N points in a space of dimensionality ``dim >= 2``.
```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -251,18 +316,38 @@ class CircleModel(BaseModel):
            0., 0., 0., 0., 0., 0., 0., 0.])
+
+    The estimation can fail when — for example — all the input or output
+    points are the same.  If this happens, you will get a circl model for which
```
```suggestion
    points are the same.  If this happens, you will get a circle model for which
```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -251,18 +316,38 @@ class CircleModel(BaseModel):
+    ``bool(model)`` is ``False``:
+
+    >>> # bool on a successful model (as above) gives True:
+    >>> bool(model)
```
I don't think we really want to encourage that pattern of explicitly converting to bool. Of course, it's correct, just that we want the spelling to be the following:

```
>>> if not model:
...     print("Estimation failed")
```

lagru replied:
+1 – the `if not model` version should clearer to many users not aware of `__bool__`.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -404,17 +517,37 @@ class EllipseModel(BaseModel):
            0., 0., 0., 0., 0., 0., 0., 0.])
+
+    The estimation can fail when — for example — all the input or output
+    points are the same.  If this happens, you will get a ellipse model for
```
```suggestion
    points are the same.  If this happens, you will get an ellipse model for
```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -680,6 +845,39 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
+
+    warn(
+        "Passing custom classes without `from_estimate` has been deprecated "
+        "since version 0.26 and will be removed in version 2.2. "
```
Version check here too.

matthew-brett replied:
Do you mean - printed version should be 2.0 here - or something else?

matthew-brett replied:
Assuming you are referring to version being 2.0 or 2.2 - and that we've resolved this in the meeting discussion.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -730,14 +928,24 @@ def ransac(
-         * ``residuals(*data)``
+        * Either:
+
+          * ``tf = model_class.from_estimate(*data)`` (``from_estimate`` class
```
Would start with ``from_estimate`` class method, and then give the description. Same the line below: ``estimate`` instance method: ...

stefanv reviewed (commented):
Looks close to me!

stefanv commented:
Rebasing on main should also avoid the pyodide failure.

matthew-brett commented on skimage/_shared/utils.py:
```diff
@@ -108,22 +137,28 @@ class change_default_value(_DecoratorBaseClass):
         Optional warning message. If None, a generic warning message
         is used.
-
+    stacklevel : int, optional
```
@lagru - how about this suggestion?

lagru replied:
`int, optional` is the convention we are using in most places. So that's fine. :+1: It's a short-hand for `int | None`. docstub is also be able to deal with that naturally.

matthew-brett commented on skimage/_shared/utils.py:
```diff
@@ -496,6 +603,12 @@ class deprecate_func(_DecoratorBaseClass):
     hint : str, optional
         A hint on how to address this deprecation,
         e.g., "Use `skimage.submodule.alternative_func` instead."
+    stacklevel : int, optional
```
@lagru - what do you think?

matthew-brett reviewed (commented):
Response to review, more to come.

matthew-brett commented:
@stefanv - I think there are now just a couple of queries remaining on your comments about versions.

matthew-brett reviewed (commented):
I think we've resolved all outstanding issues here - do y'all agree?

lagru commented on skimage/transform/_thin_plate_splines.py:
```diff
@@ -66,6 +66,31 @@ class ThinPlateSplineTransform:
+    >>> bad_tps.params  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+      ...
+    FailedEstimationAccessError: No attribute "params" for failed estimation ...
```
I really dig the example section. :ok_hand:

lagru commented on skimage/transform/_thin_plate_splines.py:
```diff
@@ -119,22 +145,36 @@ def estimate(self, src, dst):
+        tform : Self or ``FailedEstimation``
+            An instance of the transformation if the estimation succeeded.
+            Otherwise, a sentinel object will be returned and signal a failed
+            estimation. Testingn the truth value of the failed estimation
```
```suggestion
            estimation. Testing the truth value of the failed estimation
```

lagru commented on skimage/transform/_thin_plate_splines.py:
```diff
@@ -119,22 +145,36 @@ def estimate(self, src, dst):
+        msg = tf._estimate(src, dst)
+        return tf if msg is None else FailedEstimation(f'{cls.__name__}: {msg}')
+
+    def _estimate(self, src, dst):
```
```suggestion
    def _estimate(self, src, dst):
        """Try to estimate and return reason if estimation fails.""" 
```

lagru commented on skimage/transform/tests/test_thin_plate_splines.py:
```diff
@@ -10,58 +10,61 @@
 
 
 class TestThinPlateSplineTransform:
+    tform_class = ThinPlateSplineTransform
```
Curious about the reason for the class attribute here? Did you plan to inherit from this class? Otherwise I'm not really seeing a reason for the indirection. 

But fine with me and not a blocker.

lagru commented on skimage/measure/fit.py:
```diff
@@ -680,6 +848,44 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
+
+    class FromEstimated(cls):
+        @classmethod
+        def from_estimate(klass, *args, **kwargs):
```
"klass" :smile:

matthew-brett commented on skimage/transform/_geometric.py:
```diff
@@ -188,7 +188,9 @@ def _umeyama(src, dst, estimate_scale):
 
     # Eq. (40) and (43).
-    rank = np.linalg.matrix_rank(A)
+    # Matrix rank calculation from SVD (see numpy.linalg code).
```
To recap - this change is basically cosmetic - but I noticed the inefficiency of calculating matrix rank here, when we already have the SVD, from which `matrix_rank` is calculated.   Is that sufficient, or would you like more explanation?

matthew-brett commented on skimage/measure/fit.py:
```diff
@@ -730,14 +936,27 @@ def ransac(
+
+        * ``residuals(*data)``
+
+        See :class:`RansacModel` for the protocol definition.  In general,
+        ensure that `model_class` is an instance of ``RansacModel`` (this
+        checks it conforms to the protocol).
```
Sorry for my ignorance, but how would I do this.  Literally add `RansacModelProtocol` to the `__all__` variable in `measure/__init__.py`?  Won't this short-circuit lazy-loading?

matthew-brett replied:
`RansacModelProtocol` is already in `__all__` of `measure/__init__.pyi` - but maybe that's not what you mean?

matthew-brett commented on skimage/transform/_thin_plate_splines.py:
```diff
@@ -119,22 +145,36 @@ def estimate(self, src, dst):
 
         Returns
         -------
-        success: bool
-            True indicates that the estimation was successful.
+        tform : Self or ``FailedEstimation``
+            An instance of the transformation if the estimation succeeded.
+            Otherwise, a sentinel object will be returned and signal a failed
+            estimation. Testingn the truth value of the failed estimation
+            sentinel will return ``False``. E.g.
+
+            .. code-block:: python
+
+                tform = ThinPlateSplineTransform.from_estimate(...)
+                if not tform:
+                    # Handle failed estimation
 
         Notes
         -----
         The number N of source and destination points must match.
         """
+        tf = cls()
+        msg = tf._estimate(src, dst)
+        return tf if msg is None else FailedEstimation(f'{cls.__name__}: {msg}')
+
+    def _estimate(self, src, dst):
         check_nD(src, 2, arg_name="src")
         check_nD(dst, 2, arg_name="dst")
 
         if src.shape[0] < 3 or dst.shape[0] < 3:
-            msg = "Need at least 3 points in in `src` and `dst`"
-            raise ValueError(msg)
+            raise ValueError("Need at least 3 points in in `src` and `dst`")
         if src.shape != dst.shape:
-            msg = f"Shape of `src` and `dst` didn't match, {src.shape} != {dst.shape}"
-            raise ValueError(msg)
+            raise ValueError(
+                f"Shape of `src` and `dst` didn't match, {src.shape} " f"!= {dst.shape}"
+            )
```
Reverted.

matthew-brett commented on skimage/measure/fit.py:
```diff
@@ -288,13 +386,12 @@ def estimate(self, data):
+            return _warn_or_msg(
                 "Standard deviation of data is too small to estimate "
                 "circle with meaningful precision.",
-                category=RuntimeWarning,
-                stacklevel=2,
+                warn_only=warn_only,
             )
-            return False
+
```
I was attempting to strictly follow the previous behavior for the (deprecated) `estimate` method. 

For the places where I'm using `_warn_or_msg`, `estimate` previously returned `True` but gave a warning.   By returning `None` in that case, I make sure the deprecated `estimate` wrapper returns `True`, replicating the previous behavior.  Because `from_estimate` always sets `warn_only` to `False` (in the `BaseModel` implementation), it will instead always fail for these cases.   In that sense, `from_estimate` behaves differently from the old `estimate` method.  Does that make sense?

matthew-brett replied:
@lagru - is my explanation convincing?  Is there anything else I can do here?

lagru replied:
In part, but I don't think we are on the same page yet. Consider the case where a the old deprecated `estimate` is called and this branch is triggered. In this case `warn_only=True` and `None` is returned. If you look at the previous diff, however, estimate would warn and *return `None`*!  Or am I just failing to follow the logic?

stefanv commented on skimage/_shared/utils.py:
```diff
@@ -23,8 +24,23 @@
-def _count_wrappers(func):
-    """Count the number of wrappers around `func`."""
+def count_inner_wrappers(func):
+    """Count the number of inner wrappers by unpacking ``__wrapped__``.
```
Worth adding a sentence on what an "inner" wrapper is? I have no idea.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -73,9 +107,21 @@ def estimate(self, data):
-        success : bool
-            True, if model estimation succeeds.
+        model : Self or ``FailedEstimation``
+            An instance of the line model if the estimation succeeded.
+            Otherwise, a sentinel object will be returned and signal a failed
+            estimation. Testing the truth value of the failed estimation
+            sentinel will return ``False``. E.g.
```
```suggestion
            An instance of the line model if the estimation succeeded.
            Otherwise, a ``FailedEstimation`` object is returned, which evaluates
            to False. To handle failed estimations, do:
```

I think jargon like "sentinel object" here may confuse users unnecessarily.

matthew-brett replied:
Rewritten.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -251,18 +316,41 @@ class CircleModel(BaseModel):
+    points are the same.  If this happens, you will get a transform that is not
+    "truthy" - meaning that ``bool(tform)`` is ``False``:
+
+    >>> # A successfully estimated model is truthy (applying ``bool()``
+    >>> # gives ``True``):
```
```suggestion
    >>> # A successfully estimated model is truthy:
```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -404,17 +520,37 @@ class EllipseModel(BaseModel):
+    which ``bool(model)`` is ``False``:
+
+    >>> # bool on a successful model (as above) gives True:
+    >>> bool(ellipse)
```
I think we should use the `if model` pattern consistently everywhere, and not encourage `bool(model)`.

matthew-brett replied:
Yup, sorry, missed that one - changed.

stefanv commented on skimage/measure/fit.py:
```diff
@@ -644,6 +787,31 @@ def predict_xy(self, t, params=None):
+        -------
+        success : bool
+            True, if model estimation succeeds.
+
+
```
```suggestion

```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -730,14 +936,27 @@ def ransac(
+
+          * ``from_estimate`` class method returning transform instance, as in
+            ``tform = model_class.from_estimate(*data)``; the resulting
+            ``tform`` should be truthy (``bool(tform) == True``) where
```
Should we add the check we recommend here?

```
if tform: ...
```

matthew-brett replied:
Hmm - how to convey that in the docstring?   Put an embedded code example in there?   Can we assume that people prepared to make custom classes won't be much distracted by the `bool(tform)` notation here?

stefanv commented on skimage/measure/fit.py:
```diff
@@ -730,14 +936,27 @@ def ransac(
+
+          * ``from_estimate`` class method returning transform instance, as in
+            ``tform = model_class.from_estimate(*data)``; the resulting
+            ``tform`` should be truthy (``bool(tform) == True``) where
```
Should we add the check we recommend here?

```
if tform: ...
```

stefanv commented on skimage/measure/fit.py:
```diff
@@ -730,14 +936,27 @@ def ransac(
+
+        * ``residuals(*data)``
+
+        See :class:`RansacModelProtocol` for the protocol definition.  In
+        general, ensure that `model_class` is an instance of
+        ``RansacModelProtocol`` (this checks it conforms to the protocol).
```
Users who are not familiar with Protocols (and many aren't) may be confused by this.
Perhaps we can give them clearer directions: "Your model should conform to (implement all of the methods / attributes specified by) RansacModelProtocol. An easy check to see whether that is the case is `isinstance(MyModel, RansacModelProtocol)`."

I'm not sure if the above is the correct way to check; is it?

matthew-brett replied:
It is.  I've expanded the text a bit.

stefanv reviewed (commented):
I'd like to get this merged ASAP. There is a gallery failure that looks related (panorama). Otherwise, all my comments are around documentation; I'll let you decide how / whether to handle those.

matthew-brett reviewed (commented):
Right back at ya!

lagru commented on doc/examples/transform/plot_geometric.py:
```diff
@@ -98,8 +99,44 @@
+bad_tform = transform.ProjectiveTransform.from_estimate(bad_src, dst)
+bad_tform
+
+######################################################################
+# If you try and use any attributes of this failed estimation, you get
+# a ``FailedEstimationAccessError``.
+
+try:
+    bad_tform.params
+except FailedEstimationAccessError as exc:
+    print('We got the following error:')
+    print(exc)
```
```suggestion
```
Personally, I think this reads better if we skip this bit? That part of the API is only really relevant if users ignore the parts above and should be pretty "self-documenting".

matthew-brett replied:
Ok - thanks - I've moved the failed estimation discussion to its own section, as you suggested below.

lagru commented on doc/examples/transform/plot_geometric.py:
```diff
@@ -98,8 +99,44 @@
+    print(exc)
+
+######################################################################
+# In fact, our original estimation succeeded, so we can apply it, for example,
```
Hmm, what do you think about moving this whole thing into its own section at the end, explaining the API around failed estimations?

lagru commented on skimage/measure/tests/test_fit.py:
```diff
@@ -41,8 +42,12 @@ def test_line_model_nd_invalid_input():
 
-    assert not LineModelND().estimate(np.empty((1, 3)))
-    assert not LineModelND().estimate(np.empty((1, 2)))
+    tf = LineModelND.from_estimate(np.empty((1, 3)))
+    assert not tf
+    assert str(tf) == 'LineModelND: estimate under-determined'
+    tf = LineModelND.from_estimate(np.empty((1, 2)))
```
Nitpick: it shouldn't matter for this case, but might still be defensive practice to use something determined instead of `np.empty`?
```suggestion
    tf = LineModelND.from_estimate(np.ones((1, 3)))
    assert not tf
    assert str(tf) == 'LineModelND: estimate under-determined'
    tf = LineModelND.from_estimate(np.ones((1, 2)))
```

lagru commented on skimage/measure/tests/test_fit.py:
```diff
@@ -195,16 +237,23 @@ def test_circle_model_estimate_from_small_scale_data():
         dtype=np.float64,
     )
     data = CircleModel().predict_xy(angles, params=params)
```
Out of scope: but this seems like a weird API too. With the current signature `predict_xy` should be a classmethod. Alternatively, the `CircleModel`'s constructor should accept `params=`.

lagru commented on skimage/measure/tests/test_fit.py:
```diff
@@ -460,22 +536,28 @@ def test_ellipse_model_estimate_from_far_shifted_data():
+        ),
+    ):
+        tf = EllipseModel.from_estimate(data)
+        assert not tf and str(tf).endswith(msg)
```
```suggestion
        assert not tf
        assert str(tf).endswith(msg)
```

lagru commented on skimage/measure/fit.py:
```diff
@@ -680,6 +849,44 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
+def add_from_estimate(cls):
+    """Add ``from_estimate`` method  class using ``estimate`` method"""
+
+    if hasattr(cls, 'from_estimate'):
+        if not ismethod(cls.from_estimate):
+            raise TypeError(f'Class {cls} `from_estimate` must be a ' 'class method.')
+        return cls
+
+    if not hasattr(cls, 'estimate'):
+        raise TypeError(
+            f'Class {cls} must have `from_estimate` class method '
+            'or `estimate` method.'
+        )
```
I wonder if we can we replace this in favor of simple `isinstance(cls, RansacModelProtocol)`. Is the reason you are doing this because of more helpful error messages?

matthew-brett commented on skimage/measure/fit.py:
```diff
@@ -19,10 +22,42 @@ def _check_data_atleast_2D(data):
+    """Protocol for `ransac` model class."""
+
+    @classmethod
+    def from_estimate(cls, *data): ...
+
+    def residuals(self, *data): ...
```
I believe that is to match the signature of the `ransac` function - that allows you to pass multiple arrays as input data.  These data then reach the `model_class` via `model = model_class.from_estimate(*samples)` (currently L 1153 in `fit.py`) - where `*samples` are the data arrays suitably indexed (`samples = [d[spl_idxs] for d in data]`).   Then the data reaches `residuals` via `residuals = np.abs(model.residuals(*data))`.   In both cases the model has to handle multiple input arrays.

matthew-brett commented:
Failures appear unrelated.  OK to pass this ball back to you @lagru?

lagru commented on skimage/transform/tests/test_geometric.py:
```diff
@@ -840,7 +853,7 @@ def test_union_differing_types():
             ).params
         ),
-        ((t := PiecewiseAffineTransform()).estimate(SRC, DST) and t),
+        PiecewiseAffineTransform.from_estimate(SRC, DST),
```
Nice perk of the API. :smile:

lagru commented on skimage/transform/_geometric.py:
```diff
@@ -477,10 +556,9 @@ def _setup_constraint_matrix(self, src, dst):
-        except ZeroDivisionError:
+        src_matrix, src = _center_and_normalize_points(src)
+        dst_matrix, dst = _center_and_normalize_points(dst)
+        if np.any(np.isnan(src_matrix + dst_matrix)):
```
Thanks for the [commit message](https://github.com/scikit-image/scikit-image/commit/41669f8d5171d9656036d5c0ab9b88d370faa04c) which explained this change. :blush:

lagru reviewed (approved):
Finally done. Assuming the CI failures are unrelated, I think we can merge this. 

Thanks @matthew-brett for the patience!

lagru commented:
:tada:
